### PR TITLE
Add block height versioning

### PR DESCRIPTION
**Pull request by pavelisnear**
This PR introduces a way to set specific block height for a component.
If no block height specified - the latest version is being fetched.

#### A component using two instances of the same component, but the first one is with a specific block height.

```tsx
import SandboxWithBlockHeight from 'near://pavel-pagoda.near/PavelSandbox@114965395';
import SandboxLatest from 'near://pavel-pagoda.near/PavelSandbox';

export default function () {
  return (
    <div>
      <h1>Sandbox with a block height:</h1>
      <SandboxWithBlockHeight />
      <hr />
      <h1>Sandbox latest:</h1>
      <SandboxLatest />
    </div>
  );
}
```

#### Should produce two instances of the same component, but with the respected versions:
image

Fixes https://github.com/near/bos-web-engine/issues/95

### CODE REVIEW

**Review by andy-haynes — CHANGES_REQUESTED** [review]
Awesome work 😍 

Comments inline per our discussion

**Comment by mpeterdev** [comment]
I have been testing this and it seems to be working well when versions are recent enough to be serviced by a non-archival RPC node. I am working on getting us answers on how to proceed with fetching older component versions

**Comment by pavelisnear** [comment]
### New flag added to the Inspector

#### State 1: `enableBlockHeightVersioning` is set to `true`
image

#### State 2: `enableBlockHeightVersioning` is set to `false` (default) - the block height is ignored and the latest version fetched
image

#### Component code:
```tsx
import BlockHeightWithFlagTest from 'near://pavel-pagoda.near/BlockHeightWithFlagTest@115160709';
import BlockHeightWithFlagTestLatest from 'near://pavel-pagoda.near/BlockHeightWithFlagTest';

export default function () {
  return (
    <div>
      <h1>Sandbox with a block height:</h1>
      <BlockHeightWithFlagTest />
      <hr />
      <h1>Sandbox latest:</h1>
      <BlockHeightWithFlagTestLatest />
    </div>
  );
}
```

File: packages/compiler/src/compiler.ts [hunk 1]
```diff
@@ -246,7 +256,15 @@ export class ComponentCompiler {
     // wait on CSS initialization
     await this.cssParser.init();
 
-    const componentPath = componentId.split('##')[0];
+    let [componentPath] = componentId.split('##');
+    const [componentPathWithoutBlockHeight, blockHeight] =
+      componentPath.split('@');
+    if (blockHeight && !this.enableBlockHeightVersioning) {
+      console.warn(
+        `${componentPath} has a block height specified, but the "enableBlockHeightVersioning" flag is disabled. The latest version of Component will be used.`
```
Comment (andy-haynes) [inline packages/compiler/src/compiler.ts:264]:
```suggestion
        `${componentPath} has a block height specified, but the "enableBlockHeightVersioning" flag is disabled. The latest version of ${componentPathWithoutBlockHeight} will be used.`
```

File: packages/compiler/src/source.ts [hunk 2]
```diff
@@ -25,15 +26,96 @@ export async function fetchComponentSources(
...
+
+        return pathsByBlockHeight;
+      },
+      {} as Record<string, string[]>
+    );
+
+    const componentsByBlockHeightArr = Object.entries(componentsByBlockHeight);
+
+    const responsesWithBlockHeight = await Promise.all(
+      componentsByBlockHeightArr.map(async ([blockId, keys]) => {
```
Comment (andy-haynes) [inline packages/compiler/src/source.ts:57]:
```suggestion
    const responsesWithBlockHeight = await Promise.all(
      Object.entries(componentsByBlockHeight).map(async ([blockId, keys]) => {
```

File: packages/compiler/src/compiler.ts [hunk 3]
```diff
@@ -246,7 +256,15 @@ export class ComponentCompiler {
     // wait on CSS initialization
     await this.cssParser.init();
 
-    const componentPath = componentId.split('##')[0];
+    let [componentPath] = componentId.split('##');
```
Comment (andy-haynes) [inline packages/compiler/src/compiler.ts:259]: By checking here, we're only warning for the root Component. Let's add it to `this.getComponentSources` instead to catch imports at the source.

**Review by andy-haynes — APPROVED** [review]
Looks great, excellent work! I've left some feedback but it can go in a follow-up or flag removal PR

File: packages/compiler/src/source.ts (hunk 2 above)
Comment (pavelisnear) [reply packages/compiler/src/source.ts:57]: yeah, good catch!
It made sense to use `componentsByBlockHeightArr` in the previous implementation, when the block height was mapped by the `idx` (array item number). Now it is just a redundant variable declaration.

**Pull request merged**